### PR TITLE
Slice node factorization

### PR DESCRIPTION
**Pull request by Punzo**
@jcfr  first pull request related to https://github.com/Slicer/Slicer/pull/499#issuecomment-220108704

### CODE REVIEW

File: Libs/vtkAddon/vtkAddonMathUtilities.cxx [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+bool vtkAddonMathUtilities::Matrix4x4AreEqual(const vtkMatrix4x4* m1, const vtkMatrix4x4* m2, double tolerance)
+{
+  int i,j;
+  for (i = 0; i < 4; i++)
+    {
+    for (j = 0; j < 4; j++)
+      {
+      if ( m1->GetElement(i, j) - m2->GetElement(i, j) > tolerance )
+        {
+        return 0;
```
Comment (lassoan) [inline Libs/vtkAddon/vtkAddonMathUtilities.cxx]: return false

File: Libs/vtkAddon/vtkAddonMathUtilities.cxx [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        }
+      }
+    }
+    return 1;
```
Comment (lassoan) [inline Libs/vtkAddon/vtkAddonMathUtilities.cxx]: return true

File: Libs/vtkAddon/vtkAddonMathUtilities.cxx [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+//----------------------------------------------------------------------------
+void vtkAddonMathUtilities::PrintSelf(ostream& os, vtkIndent indent)
+{
+  this->Superclass::PrintSelf(os,indent);  
+}
+
+bool vtkAddonMathUtilities::Matrix4x4AreEqual(const vtkMatrix4x4* m1, const vtkMatrix4x4* m2, double tolerance)
+{
+  int i,j;
```
Comment (lassoan) [inline Libs/vtkAddon/vtkAddonMathUtilities.cxx]: move these into the for loops (int i=0...)

File: Libs/vtkAddon/vtkAddonMathUtilities.cxx [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+  for (i = 0; i < 4; i++)
+    {
+    for (j = 0; j < 4; j++)
+      {
+      if ( m1->GetElement(i, j) - m2->GetElement(i, j) > tolerance )
```
Comment (lassoan) [inline Libs/vtkAddon/vtkAddonMathUtilities.cxx]: either add fabs or check < and > as well ! ! !

File: Libs/vtkAddon/vtkAddonMathUtilities.cxx [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+  See the License for the specific language governing permissions and
+  limitations under the License.
+
+===============================================================================auto=*/
+
+#include <vtkAddonMathUtilities.h>
+#include <vtkMath.h>
+#include <vtkNew.h>
+#include <vtkObjectFactory.h>
+#include <math.h>
```
Comment (jcfr) [inline Libs/vtkAddon/vtkAddonMathUtilities.cxx]: `math.h` shouldn't be required

**Comment by lassoan** [comment]
Thanks! It's good that you could find and eliminate all those duplicate implementations.

Please add a test. The big fabs mistake underlines the importance of adding tests for even things as simple as this.

**Comment by Punzo** [comment]
yeah I'll do, thank. it was a very bad mistake!

**Comment by jcfr** [comment]
I already started to the tweak reported by @lassoan and will add a skeleton of test. Will then push an updated topic ... @Punzo  you could then implement the test and do the final tweak

**Comment by Punzo** [comment]
ok I have applied the corrections, I'll add the test. @jcfr, should the test be another pull request?

**Comment by jcfr** [comment]
Could you push a new commit on that same PR ?

**Comment by jcfr** [comment]
This PR is superseded by #503

**Pull request closed without merging**